__tests__/SqlUtils.test.ts

```diff
@@ -1,33 +1,168 @@
+            throw new mssql.ConnectionError(new Error(`Client with IP address '1.2.3.4' is not allowed to access the server.`));
         });
-        const ipAddress = await SqlUtils.detectIPAddress(new SqlConnectionConfig('Server=testServer.database.windows.net;Initial Catalog=testDB;User Id=testUser;Password=placeholder'));
+        const ipAddress = await SqlUtils.detectIPAddress(getConnectionConfig());
+
+        expect(mssqlSpy).toHaveBeenCalledTimes(1);
+        expect(ipAddress).toBe('1.2.3.4');
+    });
+
+    it('detectIPAddress should return ipaddress when connection returns AggregateError', async () => {
+        const mssqlSpy = jest.spyOn(mssql, 'connect').mockImplementation(() => {
+            const errors = new AggregateError([
+                new Error(`We don't care about this error.`),
+                new Error(`Client with IP address '1.2.3.4' is not allowed to access the server.`)
+            ])
+            throw new mssql.ConnectionError(errors);
+        });
+        const ipAddress = await SqlUtils.detectIPAddress(getConnectionConfig());
 
         expect(mssqlSpy).toHaveBeenCalledTimes(1);
         expect(ipAddress).toBe('1.2.3.4');
     });
 
     it('detectIPAddress should return empty', async () => {
-        const mssqlSpy = jest.spyOn(mssql.ConnectionPool.prototype, 'connect').mockImplementation((callback) => {
-            // Successful connections calls back with null error
-            callback(null);
+        const mssqlSpy = jest.spyOn(mssql, 'connect').mockImplementation(() => {
+            // Successful connection
+            return new mssql.ConnectionPool('');
         });
-        const ipAddress = await SqlUtils.detectIPAddress(new SqlConnectionConfig('Server=testServer.database.windows.net;Initial Catalog=testDB;User Id=testUser;Password=placeholder'));
+        const ipAddress = await SqlUtils.detectIPAddress(getConnectionConfig());
 
         expect(mssqlSpy).toHaveBeenCalledTimes(1);
         expect(ipAddress).toBe('');
     });
 
-    it('detectIPAddress should throw error', () => {
-        const mssqlSpy = jest.spyOn(mssql.ConnectionPool.prototype, 'connect');
-        expect(SqlUtils.detectIPAddress(new SqlConnectionConfig('Server=testServer.database.windows.net;Initial Catalog=testDB;User Id=testUser;Password=placeholder'))).rejects;
+    it('detectIPAddress should throw error', async () => {
+        const mssqlSpy = jest.spyOn(mssql.ConnectionPool.prototype, 'connect').mockImplementation(() => {
+            throw new mssql.ConnectionError(new Error('Failed to connect.'));
+        });
+
+        let error: Error | undefined;
+        try {
+            await SqlUtils.detectIPAddress(getConnectionConfig());
+        }
+        catch (e) {
+            error = e;
+        }
+
+        expect(error).toBeDefined();
+        expect(error!.message).toMatch('Failed to add firewall rule. Unable to detect client IP Address.');
```

Comment:
is this error message supposed to be the same as the one on line 49?

Reply:
No. The error thrown on line 49 is supposed to mock an exception thrown by SqlClient. detectIPAddress parses the error message to see if it is related to the IP address firewall rule, and if not, it will display the SqlClient error directly and then throw this "Failed to add firewall rule..." error